pkgs/misc/arm-trusted-firmware/default.nix

```diff
@@ -33,7 +33,7 @@ let
       owner = "ARM-software";
       repo = "arm-trusted-firmware";
       rev = "v${version}";
-      sha256 = "sha256-WDJMMIWZHNqxxAKeHiZDxtPjfsfQAWsbYv+0o0PiJQs=";
+      sha256 = assert version=="2.7"; "sha256-WDJMMIWZHNqxxAKeHiZDxtPjfsfQAWsbYv+0o0PiJQs=";
```

Comment:
thats really ugly. If the version should not be overwritten why don't we remove the version input in line 20?

Reply:
> why don't we remove the version input in line 20?

Yeah, I'd like to do that, but it will break any expressions that are using it.  The `buildArmTrustedFirmware` attr in `top-level.nix` is used by device-specific callers outside in packages that are not kept in nixpkgs (stuff in nixos-hardware or similar).

But I can't imagine any situation where somebody is overriding `version` and the expression is doing what they want it to do.  Do you think causing some breakage is okay in that situation?

> thats really ugly.

IMHO there are a lot of ugly things (overriding `postFetch`, the "don't use fetchpatch here" comments, etc) and footguns in nixpkgs' fetchers.  I assume most of this is waiting on `ca-derivations` before making a proper fix.

Reply:
> Yeah, I'd like to do that, but it will break any expressions that are using it.

Right now it would also break if the assert fails.

> Do you think causing some breakage is okay in that situation?

Yes

> don't use fetchpatch here

Those are required because we need to bootstrap fetchpatch somehow but adding inline asserts like this when we have a nicer alternative is not something we must do.

Reply:
> > Do you think causing some breakage is okay in that situation?
> 
> Yes

Done: 72679fb5efb77a236aada53ed54493a42d2fb42d

> > don't use fetchpatch here

> Those are required because we need to bootstrap fetchpatch somehow

Maybe this isn't the place to discuss it, but the easy fix is to simply have 

```
  fetchpatchBoot = callPackage .../fetchpatch/boot.nix ...
```

... just like `fetchurlBoot`.  The only thing it needs is perl, and `fetchpatchBoot` can use `${bootstrap-files}/bin/perl`.  Then all those comments and the footgun both go away.

Is there some obvious reason why this wouldn't work?  If not I can write a PR.

My other gripe is that the `hash` passed to `fetchpatch` should be the hash of the patch *before* it is modified, but that's a longer discussion.